### ISSUE TITLE
TimeZone support

### ISSUE THREAD

**Issue by mattjohnsonpint** [andyearnshaw/Intl.js#19]
Do you have plans already on how to shim the enhanced TimeZone support?

**Comment by caridy** [comment]
for the record, this is what we have in chrome today:

```js
new Intl.DateTimeFormat('en', {
    weekday: "long",
    timeZone: "UTC",
    timeZoneName: "short",
    month: "long",
    day: "numeric",
    year: "numeric",
    hour: "numeric",
     minute: "2-digit",
     second: "2-digit"
}).format(Date.now());
// "Wednesday, March 25, 2015, 10:08:13 PM GMT"
```

while the polyfill will output: ``'Wednesday, March 25, 2015, 10:11:54 PM'`

**Comment by longlho** [comment]
Having used to maintain [timezone-js](https://github.com/mde/timezone-js), timezone is a whole other can of worms that is both natively buggy cross-browser and confusing as well. Although it'd be great if it is supported :smile:

**Comment by mj1856** [comment]
@longlho  - [moment-timezone](http://momentjs.com/timezone/) has gotten a lot better lately. :)

**Comment by caridy** [comment]
@longlho do you want to give it a try now that the code is a little better organized? jejeje

**Comment by longlho** [comment]
lol sure :+1:  :)

**Comment by kennethaa** [comment]
This will be very useful. I'm using this polyfill for node, and I want to generate a time with timezone specific settings from uts.

@longlho, plans for implementing timezone anytime soon?

**Comment by caridy** [comment]
I'm still looking for a volunteer :)

**Comment by longlho** [comment]
I can *try* to take a look :) The core of it is the IANA tz parser which is hairy, and then mapping to this library itself.

**Comment by mj1856** [comment]
Yes, parsing is tricky, and data maintenance is a pain.

Perhaps consider a plugin model.  Expose some API in Intl.js that can be passed a function for time zone conversions.  Then give example(s) of plugging in to existing libraries (like moment-timezone).

Doing it all yourself is possible, but may prove to be more hassle than its worth.

**Comment by mj1856** [comment]
Also, note that ECMA-402/2.0 makes a lot of the time zone stuff required instead of optional.

**Comment by longlho** [comment]
yeah agree w/ @mj1856

**Comment by jpierson** [comment]
I noticed and subscribed to this issue after posting a [related question on StackOverflow](http://stackoverflow.com/questions/35279594/how-can-i-output-long-datetime-format-which-includes-timzone-with-moment-timezon). In my case I assumed that I would have to make use of a library like moment-timezone because I could find no mention of timezone support in the FormatJS or react-intl documentation. It sounds like from @caridy's comment that that there at least partial support for browsers other than Safari at the moment?

**Comment by caridy** [comment]
@jpierson that's correct. All browsers, except safari (which is under dev), have support for timezone. The polyfill doesn't as today! We need a brave sold to help with that. All the data is in CLDR, and I can help with the extraction process, and provision of the DATA, which is probably the most convoluted part, the rest should be easy ;)

**Comment by mj1856** [comment]
Actually, most browsers don't have this yet.  See http://kangax.github.io/compat-table/esintl/ (under "DateTimeFormat" - "resolvedOptions().timeZone defaults to the host environment"

**Comment by caridy** [comment]
@zbraniecki do you know about this issue in FF?

@mj1856 thanks for pointing that out, I will follow up with your co-worker @bterlson as well to see when are they planning to fix that :)

**Comment by efkan** [comment]
Gentlemen, is there any progress about this issue?

I got stuck on [`timeZone is not supported`](https://github.com/andyearnshaw/Intl.js/blob/68514268e4f6b97a9323f6be9e75bd5a1216777a/src/12.datetimeformat.js#L197-L198)

**Comment by xdissent** [comment]
We've got a branch with tz support, but it's the older spec IIRC, and it requires an external helper for tz data and conversions: https://github.com/bkon-connect/Intl.js/tree/tz  We never sent a PR because of the external requirement (although looking at it now, it's totally optional). I'd gladly update it if I could get some guidance on how the external tool should ideally be integrated, or the go-ahead to include tz data and some utils directly.

**Comment by markandey** [comment]
@caridy @xdissent 

[date-time-format-timezone](https://github.com/yahoo/date-time-format-timezone) is pretty much independent,  which can polyfill intl.js implementation too. 
I started working on similar approach what @xdissent done. Then I realized that {locale+timezone data} combination is not easy to deal with. 

Would be best if you simply direct people to use [date-time-format-timezone](https://github.com/yahoo/date-time-format-timezone) polyfill instead.

**Comment by caridy** [comment]
@markandey does it work well with Intl.js? Is it already part of polyfill.io?

**Comment by markandey** [comment]
@caridy yes they do work together.  date-time-format-timezone has test for ie9 where intl is supported by intl.js and then timezones are supported by  date-time-format-timezone . 
I am planning to make Pull request for polyfill.io, its not part of polyfill.io yet.

**Comment by caridy** [comment]
Ok @markandey, let's get that done first, cc me on the PR to ft polyfill service, and once we get it there, we can document how to use both polyfills together to support timezone.

**Comment by fgerschau** [comment]
What is the status of this issue? The documentation says that timezones are still not supported...

**Comment by longlho** [comment]
@fgerschau u can use this polyfill in conjunction with https://github.com/yahoo/date-time-format-timezone

**Comment by bramski** [comment]
Ermagerd.  How has this never been addressed?  A lot of us using intl as a polyfill for react-native.
https://github.com/facebook/react-native/issues/25595

**Comment by srl295** [comment]
@bramski does the above comment https://github.com/andyearnshaw/Intl.js/issues/19#issuecomment-405655766 solution work? If so, perhaps the docs should at least be updated to mention it?

**Comment by bramski** [comment]
It does! But this is hours of me chasing around this issue: https://github.com/moment/luxon/issues/541

**Comment by longlho** [comment]
This repo unfortunately isn't really actively maintained as browsers have caught up fairly quickly with Intl APIs.

**Comment by bramski** [comment]
Yes but the polyfill is needed on react-native Android where the Javascript runtime does not provide Intl and will never support it.

**Comment by longlho** [comment]
Sorry I don't use React Native so I'm not sure but we've received similar issues over at `react-intl`. Do this instructions work? https://github.com/react-native-community/jsc-android-buildscripts#international-variant

**Comment by bramski** [comment]
Saved my day, thanks @longlho! Yes this polyfill is not needed on react-native if you have it compile with the intl version.  Turns out the developer previous to me both included `intl` and compiled the international JSC so our react-native app broke in this really weird way when 0.59 upgraded the build configuration.  Only the time zone support broke giving me hours of headache trying to track it down when instead it should have been obvious that Intl was just gone!  That dude owes me beer....

**Comment by chrisheninger** [comment]
@longlho a combination of jsc-android-buildscripts + intl + date-time-format-timezone fixed the Android portion of my react-native app. Thank you! 🙂🙇‍♂️
